File: 022. Generate Parentheses/generateParenthesis.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    var dfs = (left,right,s) => {
        if (s.length == n*2) {
            res.push(s)
            return
        }

        if (left < n) {
            dfs(left+1,right,s+'(')
        }
        if (left > right) {
            dfs(left,right+1,s+')')
        }
    }

    res = []
    dfs(0,0,'')
    return res
};

// Runtime 117 ms / Memory 43.8 MB
// debug: https://jsfiddle.net/

File: 022. Generate Parentheses/generateParenthesis.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res{};
        DFS(0,0,"",n,res);
        return res;
    }
    void DFS(int L, int R, string s, int &n, vector<string> &res) {
        // full char , return
        if (s.length()==n*2) {
            res.push_back(s);
            return;
        }
        // can go left if left step times not full
        if (L<n) { DFS(L+1, R, s+"(", n, res); }
        // can go right if left already step
        if (L>R) { DFS(L, R+1, s+")", n, res); }
    }
};